python_modules/dagster/dagster/_model/decorator.py

```diff
@@ -276,18 +306,72 @@ def _build_checked_new_str(self) -> str:
             )
             check_calls.append(f"{name}={call_str}")
 
-        check_call_block = "        ,\n".join(check_calls)
+        check_call_block = ",\n        ".join(check_calls)
         return f"""
-def __checked_new__(cls, {kw_args}):
+def __checked_new__(cls{kw_args_str}):
+    {set_calls_str}
     return cls.__nt_new__(
-        cls,{check_call_block}
+        cls,
+        {check_call_block}
     )
 """
 
 
+def _build_defaults_new(field_set: Mapping[str, Type], defaults: Mapping[str, Any]) -> str:
+    kw_args_str, set_calls_str = _build_args_and_sets(field_set, defaults)
+    assign_str = ",\n        ".join([f"{name}={name}" for name in field_set.keys()])
+    return f"""
+def __defaults_new__(cls{kw_args_str}):
+    {set_calls_str}
+    return cls.__nt_new__(
+        cls,
+        {assign_str}
+    )
+    """
+
+
+def _build_args_and_sets(
+    field_set: Mapping[str, Type],
+    defaults: Mapping[str, Any],
+) -> Tuple[str, str]:
+    kw_args = []
+    set_calls = []
+    for arg in field_set.keys():
+        if arg in defaults:
+            default = defaults[arg]
+            if default is None:
+                kw_args.append(f"{arg} = None")
+            # dont share class instance of default empty containers
+            elif default == []:
+                kw_args.append(f"{arg} = None")
+                set_calls.append(f"{arg} = {arg} if {arg} is not None else []")
+            elif default == {}:
+                kw_args.append(f"{arg} = None")
+                set_calls.append(f"{arg} = {arg} if {arg} is not None else {'{}'}")
+            # fallback to direct reference if unknown
+            else:
+                kw_args.append(f"{arg} = {_DEFAULT_VALUES_KEY}['{arg}']")
+        else:
+            kw_args.append(arg)
+
+    kw_args_str = ""
+    if kw_args:
+        kw_args_str = f", *, {', '.join(kw_args)}"
+
+    set_calls_str = ""
+    if set_calls:
+        set_calls_str = "\n    ".join(set_calls)
+
+    return kw_args_str, set_calls_str
```

Comment:
there are getting complicated enough where some unit tests could be good, if nothing else than for documentation purposes.